### PR TITLE
Configs for themes in `Display Settings`:  hide themes options,  hide custom themes, specific list of themes

### PR DESCRIPTION
**Pull request by dmeza**
#### Summary
Add configuration values for themes in `Display Settings`:
- add config to hide/show themes section
- add config to hide/show custom themes
- add config to list only specific themes to select

Configuration options:
- Default configuration, display as platform works right now:
    "ThemeSettings": {
        "EnableThemeSelection": true,
        "AllowCustomThemes": true,
        "AllowedThemes": []
    },
screen shot 2017-08-09 at 11 22 57 pm

- Hide themes section:
    "ThemeSettings": {
        "EnableThemeSelection": false,
        "AllowCustomThemes": true,
        "AllowedThemes": []
    },
screen shot 2017-08-09 at 11 24 32 pm

- Hide custom themes:
    "ThemeSettings": {
        "EnableThemeSelection": true,
        "AllowCustomThemes": false,
        "AllowedThemes": []
    },
screen shot 2017-08-09 at 11 20 27 pm

- List only specific themes:
"ThemeSettings": {
    "EnableThemeSelection": true,
    "AllowCustomThemes": false,
    "AllowedThemes": ["mattermostDark", "windows10"]
},
screen shot 2017-08-09 at 10 23 30 pm

#### Checklist
- [ ] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [x] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

**Comment by dmeza** [comment]
@jasonblais I think this should be available only under License and I need to add system console, but wanted to get the conversation going as if you need any other changes.

**Comment by lfbrock** [comment]
@dmeza - sorry I can't tell from your comments, is there a way for an admin to set a custom theme here? Or can they only pick from the existing Mattermost themes?

**Comment by jasonblais** [comment]
Hey @dmeza, 

Thanks again for the pull request. Much appreciated. Apologize for the delay as I've been thinking through the different use cases and the value the feature adds to admins.

For use cases, I've gathered these three that would be relevant for "default" themes:
A. Wants to set a default theme (either custom or one of the built-in themes by Mattermost). Okay with end users changing this setting. [config settings 1 & 2]
B. Wants to set a default theme (either custom or one of the built-in themes by Mattermost). Doesn’t want end users to change the setting.  [config settings 1, 2 & 3]
C. Wants to set two or more default themes (either built-in or custom ones), with end users choosing between them only. [use case for Shift; supported by allowing more than one theme in `DefaultTheme`.

To support these use cases, I'm wondering if we'd be able to update the pull request to support the following three config.json settings:
 1. `EnableDefaultTheme`: Toggles whether default theme is set system-wide
     - includes an addition to diagnostics.go on whether the value is true/false
 2. `DefaultTheme`: This sets the default theme for the system. To support Shift's use case (where themes are restricted to `shift` and `monokai`), this might need to allow more than one theme vector, perhaps in a list
     - includes an addition to diagnostics.go on how many themes are set
 3. `EnableThemeSelection`: This can remain the same, no changes needed.
     - includes an addition to diagnostics.go on whether the value is true/false

The only other change would be
4. Restrict this setting to Enterprise Edition E10 & E20, similar to https://github.com/mattermost/platform/pull/7140

Please let me know if you feel these changes are beyond the scope of this PR. If so, we can discuss taking over from some of these as well.

----

The reason why I'm proposing the above changes (1-3) is that currently the PR restricts theme selection from end users. But it's not clear how much value it actually adds for admins, as they'd like to set the default theme first, and then restrict theme selection from end users.

We'll also want to add a System Console page for these settings, similar to attached, but we can do it ourselves later:

image

**Comment by jasonblais** [comment]
@dmeza Let me know if you have any questions.

Just a heads up that Mattermost will be separating the `/platform` repo into two repositories on September 6th:
 - a new repository for webapp client code, hosted at github.com/mattermost/mattermost-webapp
 - existing repository (`/platform`) renamed to `/mattermost-server`, containing the server code.

Open PRs on September 6th may need to re-submit their pull request after the separation. For instance, PRs changing the UI would need to be re-submitted to the new `mattermost-webapp` repository.

More details here: https://forum.mattermost.org/t/mattermost-separating-platform-into-two-repositories-on-september-6th/3708/3

**Comment by dmeza** [comment]
@jasonblais My questions are if `EnableDefaultTheme: true` would the user be able to changes themes? or what's the difference with the current default theme?

And if `"DefaultTheme": ["shift",  "monokai"]` is an array, then we don't need `"AllowCustomThemes": false`, or  `"AllowedThemes": ["mattermostDark", "windows10"]`?

**Comment by jasonblais** [comment]
Hey @dmeza. Originally I was thinking those wouldn't be needed, but after chatting with Lindsay we can keep them. Also, the `EnableDefaultTheme` can be omitted for now (we might need it for the System Console UI, but we're not doing it now).

So just to summarize my previous note, the only changes would be
1 - Add `DefaultTheme` setting. This sets the default theme for the system.

Default value is `mattermost`. Available values include the built-in themes (`mattermost`, `organization`, `mattermostDark` and `windowsDark`). For Shift, it could include your built-in themes like `monokai` and `shift`

2 - Restrict this setting to Enterprise Edition E10 & E20, similar to #7140
3 - Add a value to `app/diagnostics.go` for the following settings:
 - `EnableThemeSelection`
 - `AllowCustomThemes`
 - `DefaultTheme`
 - `AllowedThemes` --> This would be a count of strings in the array, not the actual values. Similar to https://github.com/mattermost/platform/blob/master/app/diagnostics.go#L262

**Comment by dmeza** [comment]
Hey @jasonblais, the changes requested are ready. The only comment is that the default value of `DefaultTheme` would be `default` since that's the name of the theme. Otherwise we would need to use the type of the theme:
`
default: {
            type: 'Mattermost',
`
https://github.com/mattermost/platform/blob/master/webapp/utils/constants.jsx#L506

**Comment by jasonblais** [comment]
Thanks @dmeza! 

Ah okay, yeah that's fine since this feature is experimental/beta. I suppose the available values are `default`, `organization`, `mattermostDark` and `windows10`?

PS: Looks like there's a build failure, can you take a look?

**Comment by dmeza** [comment]
@jasonblais thank you for checking. It's fixed now. All checks passed.

File: model/config.go [hunk 1]
```diff
@@ -129,6 +129,8 @@ const (
 	ANNOUNCEMENT_SETTINGS_DEFAULT_BANNER_COLOR      = "#f2a93b"
 	ANNOUNCEMENT_SETTINGS_DEFAULT_BANNER_TEXT_COLOR = "#333333"
 
+	TEAM_SETTINGS_DEFAULT_TEAM_TEXT = "mattermost"
```
Comment (jasonblais) [inline model/config.go]: Should this be `"default"`?

**Comment by dmeza** [comment]
@jasonblais you are right. Fixed the constant value to `TEAM_SETTINGS_DEFAULT_TEAM_TEXT = "default"`. The value of `DefaultTheme` is a string that would be any key listed on `constants. THEMES` https://github.com/mattermost/platform/blob/master/webapp/utils/constants.jsx#L504.

The logic is that if `DefaultTheme` is empty or has a key that does not exist in `constants. THEMES` it will fallback to `default` as it was before.

**Comment by jasonblais** [comment]
@dmeza I tried testing this PR but looks like the Theme settings are broken (on default config settings):

1) Go to Account Settings > Display > Theme --> Get a JavaScript error
2) Try to hit "X" to close the Account Settings modal --> Also get an error.

Errors are attached below

image

**Comment by dmeza** [comment]
@jasonblais fixes the issues with undefined values because of the unlicensed server. I tested that licensed and not licensed work. Sorry I missed those tests.

**Comment by jasonblais** [comment]
Thanks @dmeza! I did a quick smoke test, will do a bit more testing Tuesday morning. Will aim to get it merged for 4.2.

**Comment by jasonblais** [comment]
Thanks @dmeza, I tested and didn't find any major issues.

There were a couple that I noticed, but those are okay given this setting will be considered Beta. I listed them here in this ticket: https://mattermost.atlassian.net/browse/PLT-7543

**Comment by dmeza** [comment]
@jasonblais: this was rebased to the latest on master and all checks passed.

File: webapp/components/user_settings/user_settings_theme.jsx [hunk 2]
```diff
@@ -180,10 +180,11 @@ export default class ThemeSetting extends React.Component {
         }
 
         const displayCustom = this.state.type === 'custom';
+        const AllowCustomThemes = global.window.mm_config.AllowCustomThemes !== 'false';
```
Comment (jwilander) [inline webapp/components/user_settings/user_settings_theme.jsx]: Normally I'd complain that this doesn't start with a lowercase but not worth blocking merging for

File: app/diagnostics.go [hunk 3]
```diff
@@ -330,6 +331,13 @@ func trackConfig() {
 		"show_full_name":     utils.Cfg.PrivacySettings.ShowFullName,
 	})
 
+	SendDiagnostic(TRACK_CONFIG_THEME, map[string]interface{}{
+		"enable_theme_selection": *utils.Cfg.ThemeSettings.EnableThemeSelection,
+		"default_theme":          *utils.Cfg.ThemeSettings.DefaultTheme,
```
Comment (grundleborg) [inline app/diagnostics.go]: @jasonblais this is a user defined string. Should it be reported, or should it just report whether the default value has been changed or not?

Comment (jasonblais) [reply app/diagnostics.go]: @dmeza can you help update this one to use the `isdefault()` function, similar to https://github.com/mattermost/platform/blob/master/app/diagnostics.go#L205?

That way it'll check if the default is used, but not actually take the user defined string.

**Comment by jasonblais** [comment]
I think the diagnostic change makes sense. @dmeza can you help with that?

@dmeza @grundleborg The server side check would be nice, but we can implement this later given the setting is experimental. Same with System Console UI. Added them to this ticket https://mattermost.atlassian.net/browse/PLT-7543

File: app/diagnostics.go (hunk 3 above)
Comment (dmeza) [reply app/diagnostics.go]: @grundleborg @jasonblais: the user defined string is for "default_theme"?

Comment (jasonblais) [reply app/diagnostics.go]: Ah yes, that's correct. For the `DefaultTheme` setting, I think just knowing if the default was used is enough.

All other settings are fine as-is

File: webapp/components/user_settings/user_settings_theme.jsx (hunk 2 above)
Comment (dmeza) [reply webapp/components/user_settings/user_settings_theme.jsx]: Fixed

**Pull request merged**